src/main/kotlin/cacophony/semantic/FunctionAnalysis.kt

```diff
@@ -14,13 +14,28 @@ data class ParentLink(
     val used: Boolean,
 )
 
+/**
+ * Analyzed function properties
+ *
+ * @property function Function declaration
+ * @property parentLink Link to wrapping function
+ * @property variables All variables read, written, or declared in the function
+ * @property auxVariables All auxiliary variables created in the function
+ * @property staticDepth Depth of the function - how deeply nested it is
+ * @property variablesUsedInNestedFunctions Variables declared in this function that are used in nested functions
+ */
 data class AnalyzedFunction(
+    val function: Definition.FunctionDeclaration,
     val parentLink: ParentLink?,
     val variables: Set<AnalyzedVariable>,
     val auxVariables: MutableSet<Variable.AuxVariable>,
     val staticDepth: Int,
     val variablesUsedInNestedFunctions: Set<Definition>,
-)
+) {
+    fun declaredVariables() = variables.filter { it.definedIn == function }
+
+    fun outerVariables() = variables.filter { it.definedIn != function }
```

Comment:
Maybe these could be `val ... by lazy {...}` to compute it just once and only if necessary?

Reply:
snapshots don't compile with such solution ☹️